ui/desktop/src/hooks/useMessageStream.ts

```diff
@@ -0,0 +1,483 @@
+
+                switch (parsedEvent.type) {
+                  case 'Message':
+                    // Update messages with the new message
+                    currentMessages = [...currentMessages, parsedEvent.message];
+                    mutate(currentMessages, false);
+                    break;
+                  
+                  case 'Error':
+                    throw new Error(parsedEvent.error);
+                  
+                  case 'Finish':
+                    // Call onFinish with the last message if available
+                    if (onFinish && currentMessages.length > 0) {
+                      const lastMessage = currentMessages[currentMessages.length - 1];
+                      onFinish(lastMessage, parsedEvent.reason);
+                    }
+                    break;
+                }
+              } catch (e) {
+                console.error('Error parsing SSE event:', e);
+                if (onError && e instanceof Error) {
+                  onError(e);
+                }
+              }
+            }
+          }
+        }
+      } catch (e) {
+        if (e instanceof Error && e.name !== 'AbortError') {
+          console.error('Error reading SSE stream:', e);
+          if (onError) {
+            onError(e);
+          }
+        }
+      } finally {
+        reader.releaseLock();
+      }
+
+      return currentMessages;
+    },
+    [mutate, onFinish, onError]
+  );
+
+  // Send a request to the server
+  const sendRequest = useCallback(
+    async (requestMessages: Message[]) => {
+      try {
+        mutateLoading(true);
+        setError(undefined);
+
+        // Create abort controller
+        const abortController = new AbortController();
+        abortControllerRef.current = abortController;
+
+        // Log the request messages for debugging
+        console.log('Sending messages to server:', JSON.stringify(requestMessages, null, 2));
+
+        // Send request to the server
+        const response = await fetch(api, {
+          method: 'POST',
+          headers: {
+            'Content-Type': 'application/json',
+            'X-Secret-Key': getSecretKey(),
+            ...extraMetadataRef.current.headers,
+          },
+          body: JSON.stringify({
+            messages: requestMessages,
+            ...extraMetadataRef.current.body,
+          }),
+          signal: abortController.signal,
+        });
+
+        if (onResponse) {
+          await onResponse(response);
+        }
+
+        if (!response.ok) {
+          const text = await response.text();
+          throw new Error(text || `Error ${response.status}: ${response.statusText}`);
+        }
+
+        // Process the SSE stream
+        const updatedMessages = await processMessageStream(response, requestMessages);
+
+        // Auto-submit when all tool calls in the last assistant message have results
+        if (maxSteps > 1 && updatedMessages.length > requestMessages.length) {
```

Comment:
What is this part doing? Hard to grok

Reply:
This is incorrect I think, i will revisit